python/ray/serve/backend_state.py

```diff
@@ -174,53 +278,38 @@ def stop(self):
         }, (f"State must be {ReplicaState.SHOULD_STOP} or "
             f"{ReplicaState.STOPPING}, *not* {self._state}")
 
-        def drain_actor(actor_name):
-            # NOTE: the replicas may already be stopped if we failed
-            # after stopping them but before writing a checkpoint.
-            try:
-                replica = ray.get_actor(actor_name)
-            except ValueError:
-                return None
-            return replica.drain_pending_queries.remote()
-
+        self._actor.graceful_stop()
         self._state = ReplicaState.STOPPING
-        self._drain_obj_ref = drain_actor(self._actor_name)
         self._shutdown_deadline = time.time(
         ) + self._graceful_shutdown_timeout_s
 
-    def check_stopped(self):
+    def check_stopped(self) -> bool:
+        """Check if the replica has stopped. If so, transition to STOPPED.
+
+        Should handle the case where the replica has already stopped.
+        """
         if self._state == ReplicaState.STOPPED:
             return True
         assert self._state == ReplicaState.STOPPING, (
             f"State must be {ReplicaState.STOPPING}, *not* {self._state}")
 
-        try:
-            replica = ray.get_actor(self._actor_name)
-        except ValueError:
+        stopped = self._actor.check_stopped()
+        if stopped:
             self._state = ReplicaState.STOPPED
             return True
 
-        ready, _ = ray.wait([self._drain_obj_ref], timeout=0)
         timeout_passed = time.time() > self._shutdown_deadline
 
-        if len(ready) == 1 or timeout_passed:
-            if timeout_passed:
-                # Graceful period passed, kill it forcefully.
-                logger.debug(
-                    f"{self._actor_name} did not shutdown after "
-                    f"{self._graceful_shutdown_timeout_s}s, force-killing.")
+        if timeout_passed:
+            # Graceful period passed, kill it forcefully.
+            # This will be called repeatedly until the replica shuts down.
+            logger.debug(
+                f"Replica {self._replica_tag} did not shutdown after "
+                f"{self._graceful_shutdown_timeout_s}s, force-killing.")
 
-            ray.kill(replica, no_restart=True)
-            ray.util.remove_placement_group(self._placement_group)
-            self._state = ReplicaState.STOPPED
-            return True
+            self._actor.force_stop()
```

Comment:
Shouldn't this `force_stop()` be if either `timeout_passed` **or** replicas are drained?

It looks like the ActorReplicaWrapper would return `True` when all queries are drained, but the Actor would still be living

Reply:
Before the logic was 
```
if replicas drained or timeout:
   if timeout:
        log a wwarning
   kill actor
   remove placement group
```

Reply:
Ah right we do still need to force_stop even in the case that the replica drained

Reply:
I'll add a unit test for it :D

Reply:
@ijrsvt I added another method called `cleanup` and now that is called whenever the actor stops (either from forceful or graceful). Added checks for it in tests too.